File: TinyScript/java/src/main/java/translator/TAProgram.java
package translator;

import org.apache.commons.lang3.StringUtils;
import translator.symbol.StaticSymbolTable;
import translator.symbol.SymbolTable;
import translator.symbol.SymbolType;

import java.util.ArrayList;

public class TAProgram {

    private ArrayList<TAInstruction> instructions = new ArrayList<>();
    private int labelCounter = 0;
    private StaticSymbolTable staticSymbolTable = new StaticSymbolTable();

    public void add(TAInstruction code) {
        instructions.add(code);
    }
    public ArrayList<TAInstruction> getInstructions() {
        return instructions;
    }


    @Override
    public String toString() {
        var lines = new ArrayList<String>();
        for(var opcode : instructions) {
            lines.add(opcode.toString());
        }
        return StringUtils.join(lines, "\n");
    }

    public TAInstruction addLabel() {
        var label = "L" + labelCounter++;
        var taCode = new TAInstruction(TAInstructionType.LABEL, null, null, null, null);
        taCode.setArg1(label);
        instructions.add(taCode);
        return taCode;
    }


    public void setStaticSymbols(SymbolTable symbolTable) {
        for(var symbol : symbolTable.getSymbols()) {
            if(symbol.getType() == SymbolType.IMMEDIATE_SYMBOL) {
                staticSymbolTable.add(symbol);
            }
        }

        for(var child : symbolTable.getChildren()) {
            setStaticSymbols(child);
        }
    }

    public StaticSymbolTable getStaticSymbolTable(){
        return this.staticSymbolTable;
    }
}

File: TinyScript/js/src/translator/TAProgram.js
const StaticSymbolTable = require('./symbol/StaticSymbolTable')
const TAInstruction = require('./TAInstruction')
const TAInstructionType = require('./TAInstructionType')
const SymbolType = require('./symbol/SymbolType')
class TAProgram {

  constructor() {

    this.instructions = []
    this.labelCounter = 0
    this.staticSymbolTable = new StaticSymbolTable()
  }

  add(code) {
    this.instructions.push(code)
  }

  getInstructions() {
    return this.instructions
  }


  toString() {
    const lines = []
    for (let opcode of this.instructions) {
      lines.push(opcode.toString())
    }
    return lines.join("\n")
  }

  addLabel() {
    const label = "L" + this.labelCounter++
    const taCode = new TAInstruction(TAInstructionType.LABEL, null, null, null, null)
    taCode.setArg1(label)
    this.instructions.push(taCode)
    return taCode
  }


  setStaticSymbols(symbolTable) {
    for (const symbol of symbolTable.getSymbols()) {
      if (symbol.getType() == SymbolType.IMMEDIATE_SYMBOL) {
        this.staticSymbolTable.add(symbol);
      }
    }

    for (const child of symbolTable.getChildren()) {
      this.setStaticSymbols(child)
    }
  }

  getStaticSymbolTable() {
    return this.staticSymbolTable
  }
}


module.exports = TAProgram